Added support for Material Property toggles

orels1 opened the pull request:
## Feature Changes

- Added a new Toggle Type: Material Property

This allows you to pick any available shader property from a particular mesh or all of the meshes of your avatar and set it to some float value. This can be used to support UV Tile Discard that is not tied to a specific shader implementation, make quick and easy hue sliders when used with the `Slider` option of the toggle, etc.

This also utilizes GraphView's Search Window functionality to provide a Hierarchical search experience for users to easily pick what they want to edit. With some rudimentary handling of incredibly long shader property display names used by likes of Poiyomi

This should eliminate the need for creating bespoke animations when all you want is to tune some shader property on your model

## Code Changes

- Added a new StateAction: MaterialPropertyAction
- Added a new Search-Window based search GUI with tree-based search
- Wired it up to the ActionClipService

## Previews

https://github.com/VRCFury/VRCFury/assets/3798928/825bed80-c303-4a74-b497-4ded35e1a3b1

## Potential Future Work

This could potentially be expanded to support more property types, but that will most likely require expanding the action state a little bit.

This felt like a good scope for initial implementation though


## Extra Notes

I also wanted to adjust the blend shapes Select experience to use the Search Window approach to avoid endless scrolling, but decided to keep it out of this PR to make the scope manageable

## License

```
The changes made in this contribution are free and unencumbered software
released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>
```

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryActionEditor.cs:
```diff
@@ -108,6 +108,126 @@ public class VRCFuryActionDrawer : PropertyDrawer {
+                propField4.style.flexGrow = 0;
+                propField4.style.flexShrink = 0;
+                propField4.style.flexBasis = 16;
+                propField4.schedule.Execute(() => {
```
Does this automatically cancel the schedule when the visualelement is disposed or detached or something?

SenkyDragon replied:
Also, does the callback run in the main thread?

SenkyDragon replied:
We wrote basically something just like this [here](https://github.com/VRCFury/VRCFury/blob/main/com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryEditorUtils.cs#L715), but I'd be happy to replace it if this does the same.

orels1 replied:
The callback does run on main, and it does de-schedule when the element is removed.

In the particular case of what I'm trying to do there - in 2022 they added a [`TrackPropertyValue`](https://docs.unity3d.com/ScriptReference/UIElements.BindingExtensions.TrackPropertyValue.html) extension which sets up the loop for you automatically so its a bit cleaner too. But in 2019 - the `schedule.Execute().Every()` is the best approach I found

SenkyDragon replied:
We have a callback for that already called `VRCFuryEditorUtils.RefreshOnChange`. It's usually designed to re-render the entire inside of the callback, but you can also just return an empty VisualElement and then do whatever you want in the callback.

orels1 replied:
Migrated to `RefreshOnChange`

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryActionEditor.cs:
```diff
@@ -108,6 +108,126 @@ public class VRCFuryActionDrawer : PropertyDrawer {
+                    }
+                };
+
+                var affectAllMeshesProp = prop.FindPropertyRelative("affectAllMeshes");
```
Thanks for this. We should do this for Blendshape some day too.

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryActionEditor.cs:
```diff
@@ -108,6 +108,126 @@ public class VRCFuryActionDrawer : PropertyDrawer {
+                void SearchClick() {
+                    var targetWidth = row.GetFirstAncestorOfType<UnityEditor.UIElements.InspectorElement>().worldBound
+                        .width;
+                    var searchContext = new UnityEditor.Experimental.GraphView.SearchWindowContext(GUIUtility.GUIToScreenPoint(Event.current.mousePosition), targetWidth, 300);
```
Would not be surprised if this breaks during the 2022 editor upgrade. They moved a lot of stuff out of Experimental.

SenkyDragon replied:
Apparently it didn't move. Nice.

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryMaterialPropertySearchWindow.cs:
```diff
@@ -0,0 +1,96 @@
+
+namespace VF.Inspector
+{
+    public class VRCFuryMaterialPropertySearchWindow : ScriptableObject, ISearchWindowProvider
```
Doesn't need to happen in this pull, but it'd probably be nice to just make this generic and accept a list of strings / callback action when one is selected. That way we don't need to make a separate ISearchWindowProvider for each use case (since I'll probably be applying this to blendshapes).

orels1 replied:
Yeah I thought about it since I wanted to make one for the blendshapes too, but since the amount of "window-specific" code is so small, I thought it would be ok for now until this is expanded into a more proper VRCF-specific API other actions could use

orels1 replied:
Converted it to a generic SearchWindowProvider that can take an array of strings or a callback for generating search results, a callback for selecting and a userData `object` similar to other GraphView apis for any other data that might be useful

SenkyDragon commented on com.vrcfury.vrcfury/Editor/VF/Service/ActionClipService.cs:
```diff
@@ -83,6 +83,29 @@ public class ActionClipService {
+                            var defaultSource = renderer.sharedMaterials.Where(mat =>
+                                mat.HasProperty(materialPropertyAction.propertyName)).ToArray();
+                            if (defaultSource.Length != 0) {
+                                offClip.SetConstant(binding, defaultSource[0].GetFloat(materialPropertyAction.propertyName));
```
offClip is only needed if the action is forcing something special to happen to the avatar's resting state. In this case (where you're leaving it at the current value), it's not needed so it can be left off (see above in FlipbookAction).

orels1 replied:
Interesting! I'll test to see if removing it works correctly.

I initially made this feature back on version 1.150 and it wasn't picking up resting states from the scene correctly. So as soon as I saw that `offClip` is a thing when I had a moment to make this into a proper PR - I just used it without checking if its actually required.

Will circle back on that when I get a moment and update the PR accordingly

TheLastRar replied:
I've had issues in the past with material property animations not returning to the resting state (in WD on) (and instead going to the material's default value when not currently animated)
Can't remember if I had this issue with float properties or colour properties

Wonder if that is what you are seeing

SenkyDragon replied:
> I've had issues in the past with material property animations not returning to the resting state (in WD on) (and instead going to the material's default value when not currently animated)

Here specifically, offClip is not actually recorded as a true "clip" in the final avatar -- rather everything put into offClip is just baked into the avatar's resting state. So reading the current value into offClip is effectively a no-op because it just gets put right back into the avatar object and then the clip is thrown away.

If there is a case where unity does not restore a default when WD in on, a fix for that should be in a separate new builder, where we force recording of certain properties that we know unity won't restore properly.

orels1 replied:
Got a chance to look at this, seems to work find without the `offClip` 👍

SenkyDragon commented:
Thanks for this. Looks awesome, and I actually learned some thing :). Didn't realize the unity search window had an api.

orels1 commented:
I have addressed the PR feedback and I think it should be good now.
Do you accept the docs PRs? I see the website on the other repo but idk if I need to write the docs for it or if you prefer doing it yourself

SenkyDragon commented:
I hate writing docs. If you'd like to contribute them, PLEASE FEEL FREE. Otherwise they probably just won't be updated 😅. I'll merge this as soon as I get back around to handing PRs, I don't recall there being anything that should block this though.
